File: libs/east-diagnostics/src/rules/no-build-time-clock.ts
/**
 * Copyright (c) 2025 Elara AI Pty Ltd
 * Dual-licensed under AGPL-3.0 and commercial license. See LICENSE for details.
 */
import type * as ts from "typescript";
import type { EastRule, TsModule } from "../types.js";
import { importsEastPackage } from "../east-source.js";

const NAME = "no-build-time-clock";
const CODE = 990027;

/** Is `node` lexically inside any function-like scope? Module-scope clock reads
 * are the abuse; a platform `.implement((…) => Date.now())` body runs at
 * RUNTIME and is fine. */
function insideFunction(node: ts.Node, t: TsModule): boolean {
  let cur: ts.Node | undefined = node.parent;
  while (cur !== undefined) {
    if (
      t.isArrowFunction(cur) ||
      t.isFunctionExpression(cur) ||
      t.isFunctionDeclaration(cur) ||
      t.isMethodDeclaration(cur) ||
      t.isConstructorDeclaration(cur) ||
      t.isGetAccessorDeclaration(cur) ||
      t.isSetAccessorDeclaration(cur)
    ) {
      return true;
    }
    cur = cur.parent;
  }
  return false;
}

// East/e3 source is compiled and DEPLOYED; reading the clock at module
// evaluation (`Date.now()`, argless `new Date()`) freezes the BUILD moment into
// the program — a seed of `new Date(Date.now() - 2h)` means "2 hours before the
// deploy", drifting meaningless afterwards. Author constant datetimes, or read
// time at runtime inside a task via the Time platform. (Complements
// `no-compile-time-seed-data`, which only inspects `e3.input` seed arguments.)
export const noBuildTimeClock: EastRule = {
  name: NAME,
  code: CODE,
  description:
    "Flag Date.now() / argless new Date() at module scope of East/e3 source — the build clock gets baked into the deployed program.",
  check(node, ctx) {
    const t = ctx.ts;
    let target: ts.Node | undefined;
    if (
      t.isCallExpression(node) &&
      t.isPropertyAccessExpression(node.expression) &&
      t.isIdentifier(node.expression.expression) &&
      node.expression.expression.text === "Date" &&
      node.expression.name.text === "now"
    ) {
      target = node;
    } else if (
      t.isNewExpression(node) &&
      t.isIdentifier(node.expression) &&
      node.expression.text === "Date" &&
      (node.arguments === undefined || node.arguments.length === 0)
    ) {
      target = node;
    }
    if (target === undefined) return;
    if (!importsEastPackage(ctx.sourceFile, t)) return;
    if (insideFunction(target, t)) return;

    const sf = ctx.sourceFile;
    const start = target.getStart(sf);
    ctx.report({
      ruleName: NAME,
      code: CODE,
      start,
      length: target.getEnd() - start,
      messageText:
        "Module-scope clock read in East/e3 source — this bakes the BUILD moment into the deployed program. Author a constant datetime (`new Date(\"2026-06-30T07:00:00Z\")`), or read time at runtime inside a task (the `Time` platform).",
      category: "warning",
    });
  },
};

File: libs/east-py/packages/east-py/east/diagnostics/rules/no_build_time_clock.py
#
# Copyright (c) 2025 Elara AI Pty Ltd
# Licensed under the Business Source License 1.1. See LICENSE.md for details.
#
"""``no-build-time-clock``: reading the clock at module scope freezes the
BUILD moment into the program. East source is compiled and deployed; a
constant computed as "two hours ago" means two hours before the deploy, and
means something different every day after it. Author the datetime, or read
the clock at RUNTIME — inside a platform function, or ``east_py_std``'s
``time_now()`` in a body — so a read inside a ``def`` is never flagged, and
neither is a script's ``if __name__ == "__main__":`` block, which an import
never runs. The TypeScript rule of the same name.
"""

from __future__ import annotations

import ast

from east.diagnostics.scope import import_time_nodes
from east.diagnostics.types import Body, Context

MESSAGE = ("a module-scope clock read bakes the BUILD moment into the deployed program — author "
           "the datetime as a constant, or read the clock at runtime: inside a platform function, "
           "or east_py_std's time_now() in a body")

#: ``datetime.now()`` / ``datetime.utcnow()`` / ``datetime.today()``
_DATETIME_READS = frozenset({"now", "utcnow", "today"})
#: ``time.time()`` and friends — the module-level clock functions
_TIME_READS = frozenset({"time", "time_ns", "monotonic", "monotonic_ns", "perf_counter"})


class NoBuildTimeClock:
    name = "no-build-time-clock"
    code = 19
    category = "warning"
    description = ("No clock read at module scope of East source — it bakes the build moment into "
                   "the deployed program.")

    def check(self, body: Body, ctx: Context) -> None:
        # A read inside any callable runs at runtime; only module scope bakes in.
        del body, ctx

    def check_module(self, ctx: Context) -> None:
        for node in import_time_nodes(ctx):
            if isinstance(node, ast.Call) and _is_clock_read(node):
                ctx.report(node, self, MESSAGE)


def _is_clock_read(node: ast.Call) -> bool:
    func = node.func
    if not isinstance(func, ast.Attribute):
        return False
    if func.attr in _DATETIME_READS:
        root = func.value
        name = root.attr if isinstance(root, ast.Attribute) else getattr(root, "id", "")
        return name in ("datetime", "date")
    return func.attr in _TIME_READS and isinstance(func.value, ast.Name) and func.value.id == "time"
